File: Day - 1/code/04_Implemented_trie_1.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isEndOfWord;

    TrieNode() : isEndOfWord(false) {}
};

class Trie {
private:
    TrieNode* root;

public:
    Trie() {
        root = new TrieNode();
    }

    void insert(const string& word) {
        TrieNode* current = root;
        for (char ch : word) {
            if (!current->children.count(ch)) {
                current->children[ch] = new TrieNode();
            }
            current = current->children[ch];
        }
        current->isEndOfWord = true;
    }

    bool search(const string& word) {
        TrieNode* current = root;
        for (char ch : word) {
            if (!current->children.count(ch)) {
                return false;
            }
            current = current->children[ch];
        }
        return current->isEndOfWord;
    }

    bool startsWith(const string& prefix) {
        TrieNode* current = root;
        for (char ch : prefix) {
            if (!current->children.count(ch)) {
                return false;
            }
            current = current->children[ch];
        }
        return true;
    }
};

int main() {
    Trie trie;
    trie.insert("apple");
    trie.insert("app");
    trie.insert("banana");

    cout << boolalpha;  // print true/false instead of 1/0
    cout << trie.search("apple") << endl;    // true
    cout << trie.search("app") << endl;      // true
    cout << trie.search("appl") << endl;     // false
    cout << trie.search("banana") << endl;   // true
    cout << trie.startsWith("app") << endl;  // true
    cout << trie.startsWith("ban") << endl;  // true
    cout << trie.startsWith("bat") << endl;  // false

    return 0;
}

File: Day - 1/code/04_Implemented_trie_1.py
# TrieNode class
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False


# Trie class
class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        current.is_end_of_word = True

    def search(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        return current.is_end_of_word

    def starts_with(self, prefix):
        current = self.root
        for char in prefix:
            if char not in current.children:
                return False
            current = current.children[char]
        return True



if __name__ == "__main__":
    trie = Trie()
    trie.insert("apple")
    trie.insert("app")
    trie.insert("banana")

    print(trie.search("apple"))    # True
    print(trie.search("app"))      # True
    print(trie.search("appl"))     # False
    print(trie.search("banana"))   # True
    print(trie.starts_with("app")) # True
    print(trie.starts_with("ban")) # True
    print(trie.starts_with("bat")) # False
